### DISCUSSION TITLE
Connection to OPCUA

### DISCUSSION THREAD

**Discussion by Zmeypp** [FreeOpcUa/python-opcua#1523] [Q&A]
Hello, I have a problem about the connection to OPC UA, if I take the architecture into ThingWorx Industrial Connectivity, we have something like this :

Connectivity (opc.tcp://192.168.130.240:49320)
--------->Device 1
--------->Device 2 (opc.tcp://192.168.120.200:51310/CogentDataHub/DataAccess)

(Connectivity contain Device 1 and 2, like folders for example)

I can connect with Connectivity without any problem
But when i try to connect with Device 2, with the same configuration for the cert, i got this error : 

`asyncua.ua.uaerrors._auto.BadCertificateUriInvalid: "The URI specified in the ApplicationDescription does not match the URI in the certificate."(BadCertificateUriInvalid)`

Here is the cert.conf : 

```[ req ] 
default_bits = 2048 
default_md = sha256 
distinguished_name = subject 
req_extensions = req_ext 
x509_extensions = req_ext 
string_mask = utf8only 
prompt = no 
[ req_ext ] 
basicConstraints = CA:FALSE 
nsCertType = client, server 
keyUsage = nonRepudiation, digitalSignature, keyEncipherment, dataEncipherment, keyCertSign 
extendedKeyUsage= serverAuth, clientAuth 
nsComment = "OpenSSL Tutorial for IRC5 OPC-UA" 
subjectKeyIdentifier=hash 
authorityKeyIdentifier=keyid,issuer 
subjectAltName = URI:urn:freeopcua:client,IP: 192.168.130.240
[ subject ] 
countryName = SE 
stateOrProvinceName = VG 
localityName = VG 
organizationName = ABB
commonName = opcUA
```

And here is my code : 

```
import asyncio
from asyncua import Client

async def connect_to_opcua_server():
    url = "opc.tcp://192.168.120.200:51310/CogentDataHub/DataAccess"
    username = "user"
    password = "password"
    policy = "Basic256Sha256"
    mode = "SignAndEncrypt"
    certificate_path = "certificate.pem"
    private_key_path = "private_key.pem"

    # create client opc ua
    client = Client(url)
    # security
    security_string = f"{policy},{mode},{certificate_path},{private_key_path}"
    
    # define security by using set_security_string
    await client.set_security_string(security_string)
    
    # authentication
    client.set_user(username)
    client.set_password(password)
    
    
    # connect to opc
    await client.connect()
    await client.create_session()

    print("connected !\n")
    
asyncio.run(connect_to_opcua_server())
```

Keep in mind that all of this is working for "opc.tcp://192.168.130.240:49320", I really don't know why this doesn't work for Device 2 :-/

Thank for your responses

**Comment by AndreasHeine** [comment]
your client application uri does not match the uri in your cert!

urn:DESKTOP-A1D6NNH:client 
!=
urn:DESKTOP-A1D6NNH:Cogent_DataHub

**Reply by Zmeypp** [reply]
I just tried with 

client = Client(url)
client.application_uri = "urn:DESKTOP-A1D6NNH:Cogent_DataHub"

and 

subjectAltName = URI:urn:DESKTOP-A1D6NNH:Cogent_DataHub,DNS:DESKTOP-A1D6NNH

and it doesn't work : 
asyncua.ua.uaerrors._auto.BadCertificateUriInvalid: "The URI specified in the ApplicationDescription does not match the URI in the certificate."(BadCertificateUriInvalid)

**Reply by schroeder-** [reply]
First you are on the wrong project. This is to old non async library, that isn't supported anymore.
Second as I told you on Stackoverflow, you should get in contact with the server developer and ask for help or extendend diagnostics. I don't think anyone can help you here, because we don't access to the server to test against and your code and certicate are look good.
